### PR TITLE
Redirects for renamed repos

### PR DESCRIPTION
**Pull request by ethantkoenig**
Adds a new table (`repo_redirect`) to track what previous names should be redirected to.

### CODE REVIEW

**Comment by lunny** [comment]
But if A -> B, but C -> A, then A will never be visit?

**Comment by ethantkoenig** [comment]
@lunny Good point, there are a lot of bugs with reusing previously-used repo names. I'll work on finding a solution.

**Comment by ethantkoenig** [comment]
@lunny I've added changes so that whenever a repo is created or renamed, any redirects *from* the new name are deleted. I believe this should fix the problem.

File: models/repo_redirect.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		return 0, ErrRepoRedirectNotExist{OwnerID: ownerID, RepoName: repoName}
+	}
+	return redirect.RedirectRepoID, nil
+}
+
+// NewRepoRedirect create a new repo redirect
+func NewRepoRedirect(ownerID, repoID int64, oldRepoName, newRepoName string) error {
+	oldRepoName = strings.ToLower(oldRepoName)
+	newRepoName = strings.ToLower(newRepoName)
+	if err := deleteRepoRedirect(x, ownerID, newRepoName); err != nil {
```
Comment (lunny) [inline models/repo_redirect.go]: I would like you use a database transaction to do that.

**Review by lunny — CHANGES_REQUESTED** [review]

**Review by lunny — APPROVED** [review]

File: models/repo_redirect.go (hunk 1 above)
Comment (lunny) [reply models/repo_redirect.go]: ^

Comment (ethantkoenig) [reply models/repo_redirect.go]: Added :grinning:

File: templates/repo/settings/options.tmpl [hunk 2]
```diff
@@ -14,7 +14,7 @@
 						{{.CsrfTokenHtml}}
 						<input type="hidden" name="action" value="update">
 						<div class="required field {{if .Err_RepoName}}error{{end}}">
-							<label for="repo_name">{{.i18n.Tr "repo.repo_name"}}<span class="text red hide" id="repo-name-change-prompt"> {{.i18n.Tr "repo.settings.change_reponame_prompt"}}</span></label>
+							<label for="repo_name">{{.i18n.Tr "repo.repo_name"}}</label>
```
Comment (lunny) [inline templates/repo/settings/options.tmpl:17]: why remove the prompt?

**Comment by lunny** [comment]
Otherwise LGTM

File: templates/repo/settings/options.tmpl (hunk 2 above)
Comment (ethantkoenig) [reply templates/repo/settings/options.tmpl:17]: The prompt warns that "This change will affect how links relate to the repository". Once we have redirects, what the prompt says will no longer be true.

Comment (lunny) [reply templates/repo/settings/options.tmpl:17]: OK

**Comment by bkcsoft** [comment]
If I remember correcly, the way github does it is to redirect for X period of time, during that time no new repo can be created with that name. Sounds good?

**Comment by ethantkoenig** [comment]
@bkcsoft Github lets you create a new repo with the old name immediately; I just checked. Perhaps it used to be as you described, but not anymore?

I also don't know about whether Github redirects expire. I haven't found any sources that explicitly say one way or another, and I can't easily test for myself :smile:

**Comment by lunny** [comment]
conflicted

**Review by appleboy — APPROVED** [review]

**Comment by appleboy** [comment]
LGTM

**Comment by lunny** [comment]
resolved #806

**Pull request merged**